### PR TITLE
[Networking] Handling iHave broken promises (and part-2 of overpromising) spams

### PR DESCRIPTION
**Pull request by yhassanzadeh13**
This PR introduces changes to our GossipSub parameter configurations to effectively address and mitigate iHave Broken Promise attacks. An iHave Broken Promise attack is characterized by a peer advertising an iHave (indicating they have a certain message), but failing to respond to the iWant requests for that message.

We have fine-tuned the GossipSub parameters, notably `defaultBehaviourPenaltyThreshold`, `defaultBehaviourPenaltyWeight`, and `defaultBehaviourPenaltyDecay` to penalize such behavior more appropriately.

The `defaultBehaviourPenaltyThreshold` is set to `10`, meaning we at most tolerate `10` RPCs containing iHave Broken Promises. The `defaultBehaviourPenaltyWeight` applies a penalty when a peer's misbehavior exceeds the threshold. The penalty is applied to the square of the difference between the misbehavior counter and the threshold, to provide a substantial disincentive for malicious peers. If a peer misbehaves more than `sqrt(2) * 10` times the threshold, they will be graylisted or disconnected.

The `defaultBehaviourPenaltyDecay` is set to `0.9`, causing the misbehavior counter to decay by `10%` per decay interval. This means that peers have a chance to recover from their previous misbehavior and incentivizes improved behavior in the future.

### CODE REVIEW

**Review by kc1116 — APPROVED** [review]
🚀

File: insecure/integration/functional/test/gossipsub/scoring/ihave_spam_test.go [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+// TestGossipSubIHaveBrokenPromises_Below_Threshold tests that as long as the spammer stays below the ihave spam thresholds, it is not caught and
+// penalized by the victim node.
+// The thresholds are:
+// Maximum messages that include iHave per heartbeat is: 10 (gossipsub parameter).
+// Threshold for broken promises of iHave per heartbeat is: 10 (Flow-specific) parameter. It means that GossipSub samples one iHave id out of the
+// entire RPC and if that iHave id is not eventually delivered within 3 seconds (gossipsub parameter), then the promise is considered broken. We set
+// this threshold to 10 meaning that the first 10 broken promises are ignored. This is to allow for some network churn.
+// Also, per hearbeat (i.e., decay interval), the spammer is allowed to send at most 5000 ihave messages (gossip sub parameter) on aggregate, and
+// excess messages are dropped (without being counted as broken promises).
+func TestGossipSubIHaveBrokenPromises_Below_Threshold(t *testing.T) {
```
Comment (peterargue) [inline insecure/integration/functional/test/gossipsub/scoring/ihave_spam_test.go:34]: can this be run with `t.Parallel()`? same question for all the other tests

**Review by peterargue — APPROVED** [review]
Looks good. added one comment about parallelizing the tests since they seem to have long waits. long term we should considering adding some functionality to mock the time interface to make these tests more flexible and less likely to be flaky due to timing.

File: insecure/integration/functional/test/gossipsub/scoring/ihave_spam_test.go (hunk 1 above)
Comment (yhassanzadeh13) [reply insecure/integration/functional/test/gossipsub/scoring/ihave_spam_test.go:34]: https://github.com/onflow/flow-go/pull/4566/commits/8fbf9a4d67cc87973da518ffddeceba4fb9d1eea

Comment (yhassanzadeh13) [reply insecure/integration/functional/test/gossipsub/scoring/ihave_spam_test.go:34]: https://github.com/onflow/flow-go/pull/4566/commits/9f2378054a24e0785479b2ceba24c991028f69ff

**Comment by yhassanzadeh13** [comment]
bors merge

**Pull request merged**